File: NapakalakiRuby/lib/specific_bad_consequence.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative 'bad_consequence'

module NapakalakiGame
  class SpecificBadConsequence < BadConsequence
    def initialize(t, l, v, h)
      bc = super(t, l)
      @specificVisibleTreasures = v
      @specificHiddenTreasures = h
    end
  
    def getSpecificHiddenTreasures
      @specificHiddenTreasures
    end
  
    def getSpecificVisibleTreasures
      @specificVisibleTreasures
    end
    
    def isEmpty
      vacio=false
      if @specificVisibleTreasures.empty? && @specificHiddenTreasures.empty? then
        vacio = true
      end
      vacio
    end
  
    def substractVisibleTreasure(t)
      indice = @specificVisibleTreasures.find_index(t.getType)
      if indice != nil then
        @specificVisibleTreasures.delete_at(indice)
      end
    end
  
    def substractHiddenTreasure(t)
      indice = @specificHiddenTreasures.find_index(t.getType)
      if indice != nil then
        @specificHiddenTreasures.delete_at(indice)
      end
    end
  
    def adjustToFitTreasureLists(v, h)
      visible = Array.new
      copy_Visible = Array.new(@specificVisibleTreasures)
      v.each do |t|
        tipo = t.getType
        fin = false
        i = 0
        while i <  copy_Visible.size && !fin
        if copy_Visible.at(i) == tipo then
          visible << copy_Visible.at(i)
          copy_Visible.delete_at(i)
          fin = true
        end
        i += 1
        end
      end
      hidden = Array.new
      copy_Hidden = Array.new(@specificHiddenTreasures)
      h.each do |t|
        tipo = t.getType
        fin = false
        i = 0
        while i <  copy_Hidden.size && !fin
        if copy_Hidden.at(i) == tipo then
          hidden << copy_Hidden.at(i)
          copy_Hidden.delete_at(i)
          fin = true
        end
        i += 1
        end
      end
      bc = SpecificBadConsequence.new(@text, @levels, visible, hidden)
      bc
    end 

    def to_s
      cadena = super
      cadena += "\nSpecific_Visible_Treasures_Down = #{@specificVisibleTreasures} / Specific_Hidden_Treasures_Down = #{@specificHiddenTreasures}" 
      cadena
    end
    
    public_class_method :new
  end
end

File: NapakalakiJava/src/NapakalakiGame/SpecificBadConsequence.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.ArrayList;

/**
 *
 * @author danibolanos & jomabose
 */
public class SpecificBadConsequence extends BadConsequence{
        private ArrayList<TreasureKind> specificVisibleTreasures;
        private ArrayList<TreasureKind> specificHiddenTreasures;
    
        public SpecificBadConsequence(String t, int l, ArrayList<TreasureKind> v,
        ArrayList<TreasureKind> h){
            super(t, l);
            specificVisibleTreasures = v;
            specificHiddenTreasures = h;
        }
        public ArrayList getSpecificHiddenTreasures(){
            return specificHiddenTreasures;
        }
    
        public ArrayList getSpecificVisibleTreasures(){
            return specificVisibleTreasures;
        }
       @Override
        public boolean isEmpty(){
            boolean vacio=false;
            if(getSpecificVisibleTreasures().isEmpty() && getSpecificHiddenTreasures().isEmpty())
                vacio = true;
            return vacio;
       }
       @Override
        public void substractVisibleTreasure(Treasure t){
            boolean quitado=false;
            TreasureKind tipo=t.getType();
            for(int i=0; i<getSpecificVisibleTreasures().size() && !quitado;i++){
                if(getSpecificVisibleTreasures().get(i)==tipo){
                    getSpecificVisibleTreasures().remove(i);
                    quitado=true;
                }
            }
        }
       @Override
        public void substractHiddenTreasure(Treasure t){
            boolean quitado=false;
            TreasureKind tipo=t.getType();
            for(int i=0; i<getSpecificHiddenTreasures().size() && !quitado;i++){
                if(getSpecificHiddenTreasures().get(i)==tipo){
                    getSpecificHiddenTreasures().remove(i);
                    quitado=true;
                }
            }
        }
        @Override
        public BadConsequence adjustToFitTreasureLists(ArrayList<Treasure> v, ArrayList<Treasure> h){
            ArrayList<TreasureKind> visible = new ArrayList();
            ArrayList<TreasureKind> specificVisible = new ArrayList(specificVisibleTreasures);
            for(Treasure t:v){
                TreasureKind tipo = t.getType();
                boolean fin=false;
                for(int i=0; i<specificVisible.size() && !fin; i++)
                    if(specificVisible.get(i)==tipo){
                        visible.add(specificVisible.get(i));
                        specificVisible.remove(i);
                        fin = true;
                    }
            }
            ArrayList<TreasureKind> hidden = new ArrayList();
            ArrayList<TreasureKind> specificHidden = new ArrayList();
            for(Treasure t:h){
                TreasureKind tipo = t.getType();
                boolean fin=false;
                for(int i=0; i<specificHidden.size() && !fin; i++)
                    if(specificHidden.get(i)==tipo){
                        hidden.add(specificHidden.get(i));
                        specificHidden.remove(i);
                        fin = true;
                    }
            }
            specificVisible= visible;
            specificHidden = hidden;
            BadConsequence bc = new SpecificBadConsequence(text, levels, specificVisible, specificHidden);
            return bc;
        }
        @Override
        public String toString(){
           String cadena = super.toString();
           cadena += "\nSpecific_Visible_Treasures_Down = " + getSpecificVisibleTreasures()
           + " / Specific_Hidden_Treasures_Down = " + getSpecificHiddenTreasures();
     
           return cadena;
        }
}
